README.md
# lddwrap
Python3 wrapper for ldd linux command


CHANGELOG.rst
1.2.2
=====
* Added support for paths with spaces (#26)

1.2.1
=====
* Removed support for Python 3.5 (#23)
* Improved support of statically linked libraries (#22)

1.2.0
=====
* Ignore not-indented ldd output (#14)
* Added handling of static libraries (#13)

1.1.0
=====
* Added ``--sorted`` command-line option
* Refactored for easier and more thorough testing

1.0.2
=====
* Added support for Python 3.7 and 3.8

1.0.1
=====
* Fixed license badge in the Readme

1.0.0
=====
* Initial version


tests/__init__.py
"""Test lddwrap."""

import os
import shlex
import tempfile
import textwrap
from typing import Optional


class MockLdd:
    """Manage context for a mock ldd script."""

    def __init__(self, out: str, out_unused: str) -> None:
        self._tmpdir = None  # type: Optional[tempfile.TemporaryDirectory]
        self.out = out
        self.out_unused = out_unused
        self._old_path = None  # type: Optional[str]

    def __enter__(self) -> None:
        self._tmpdir = tempfile.TemporaryDirectory()

        # Create a mock bach script called ``ldd`` based on
        # https://github.com/lattera/glibc/blob/master/elf/ldd.bash.in
        script = textwrap.dedent('''\
        #!/usr/bin/env bash
        if [ $# -eq 1 ]; then
            echo {out}
        elif [ "$1" = "--unused" ]; then
            echo {out_unused}

            # Return code 1 implies at least one unused dependency.
            exit 1
        else
            >&2 echo '$0 is: ' $0
            >&2 echo '$1 is: ' $1
            >&2 echo '$2 is: ' $2
            >&2 echo "Unhandled command line arguments (count: $#): $@"
            exit 1984
        fi
        '''.format(
            out=shlex.quote(self.out), out_unused=shlex.quote(self.out_unused)))

        pth = os.path.join(self._tmpdir.name, 'ldd')
        with open(pth, 'wt') as fid:
            fid.write(script)

        os.chmod(pth, 0o700)

        self._old_path = os.environ.get('PATH', None)

        os.environ['PATH'] = (self._tmpdir.name if self._old_path is None else
                              self._tmpdir.name + os.pathsep + self._old_path)

    def __exit__(self, exc_type, exc_val, exc_tb):
        self._tmpdir.cleanup()
        if self._old_path is None:
            del os.environ['PATH']
        else:
            os.environ['PATH'] = self._old_path


tests/test_ldd.py
#!/usr/bin/env python
"""Test lddwrap."""
# pylint: disable=missing-docstring,too-many-public-methods
import pathlib
import shutil
import tempfile
import unittest
from typing import Any, List, Optional

import lddwrap

import tests

# Some distros like NixOS does not have binaries on /bin.
# Instead of hardcoding them, try to get them from PATH
# using shutil.which function.
DIR = shutil.which("dir") or "/bin/dir"
PWD = shutil.which("pwd") or "/bin/pwd"


class DependencyDiff:
    """Represent a different between two dependencies."""

    def __init__(self, attribute: str, ours: Any, theirs: Any) -> None:
        self.attribute = attribute
        self.ours = ours
        self.theirs = theirs

    def __repr__(self) -> str:
        return "DependencyDiff(attribute={!r}, ours={!r}, theirs={!r})".format(
            self.attribute, self.ours, self.theirs)


def diff_dependencies(ours: lddwrap.Dependency,
                      theirs: lddwrap.Dependency) -> List[DependencyDiff]:
    """
    Compare two dependencies and give their differences in a list.

    An empty list means no difference.
    """
    keys = sorted(ours.__dict__.keys())
    assert keys == sorted(theirs.__dict__.keys())

    result = []  # type: List[DependencyDiff]
    for key in keys:
        if ours.__dict__[key] != theirs.__dict__[key]:
            result.append(
                DependencyDiff(
                    attribute=key,
                    ours=ours.__dict__[key],
                    theirs=theirs.__dict__[key]))

    return result


class TestParseOutputWithoutUnused(unittest.TestCase):
    def test_parse_line(self):
        # yapf: disable
        lines = [
            "libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 "
            "(0x00007f9a19d8a000)",
            "/lib64/ld-linux-x86-64.so.2 (0x00007f9a1a329000)",
            "libboost_program_options.so.1.62.0 => not found",
            "linux-vdso.so.1 =>  (0x00007ffd7c7fd000)",
            "libopencv_stitching.so.3.3 => not found",
            "libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 "
            "(0x00007f4b78462000)",
            "libz.so.1 => not found",
            "../build/debug/libextstr.so => not found",
            "/home/user/lib/liblmdb.so => not found",
            "/home/u s e r/lib/liblmdb.so => not found",
            "UnityPlayer.so => /home/user/games/q u d/./UnityPlayer.so "
            "(0x00007f90f290f000)"
        ]
        # yapf: enable

        expected_deps = [
            lddwrap.Dependency(
                soname="libstdc++.so.6",
                path=pathlib.Path("/usr/lib/x86_64-linux-gnu/libstdc++.so.6"),
                found=True,
                mem_address="0x00007f9a19d8a000",
                unused=None),
            lddwrap.Dependency(
                soname=None,
                path=pathlib.Path("/lib64/ld-linux-x86-64.so.2"),
                found=True,
                mem_address="0x00007f9a1a329000",
                unused=None),
            lddwrap.Dependency(
                soname="libboost_program_options.so.1.62.0",
                path=None,
                found=False,
                mem_address=None,
                unused=None),
            lddwrap.Dependency(
                soname="linux-vdso.so.1",
                path=None,
                found=True,
                mem_address="0x00007ffd7c7fd000",
                unused=None),
            lddwrap.Dependency(
                soname="libopencv_stitching.so.3.3",
                path=None,
                found=False,
                mem_address=None,
                unused=None),
            lddwrap.Dependency(
                soname="libstdc++.so.6",
                path=pathlib.Path("/usr/lib/x86_64-linux-gnu/libstdc++.so.6"),
                found=True,
                mem_address="0x00007f4b78462000",
                unused=None),
            lddwrap.Dependency(
                soname="libz.so.1", path=None, found=False, mem_address=None),
            lddwrap.Dependency(
                soname=None,
                path=pathlib.Path("../build/debug/libextstr.so"),
                found=False,
                mem_address=None,
                unused=None),
            lddwrap.Dependency(
                soname=None,
                path=pathlib.Path("/home/user/lib/liblmdb.so"),
                found=False,
                mem_address=None,
                unused=None),
            lddwrap.Dependency(
                soname=None,
                path=pathlib.Path("/home/u s e r/lib/liblmdb.so"),
                found=False,
                mem_address=None,
                unused=None),
            lddwrap.Dependency(
                soname="UnityPlayer.so",
                path=pathlib.Path("/home/user/games/q u d/./UnityPlayer.so"),
                found=True,
                mem_address="0x00007f90f290f000",
                unused=None),
        ]

        for i, line in enumerate(lines):
            # pylint: disable=protected-access
            dep = lddwrap._parse_line(line=line)

            self.assertListEqual(
                [], diff_dependencies(ours=dep, theirs=expected_deps[i]),
                "Incorrect dependency read from the line {} {!r}".format(
                    i + 1, line))

    def test_parse_wrong_line(self):
        # ``parse_line`` raises a RuntimeError when it receives an unexpected
        # structured line
        run_err = None  # type: Optional[RuntimeError]
        line = "\tsome wrong data which does not make sense"
        try:
            lddwrap._parse_line(line=line)  # pylint: disable=protected-access
        except RuntimeError as err:
            run_err = err

        self.assertIsNotNone(run_err)
        self.assertTrue(str(run_err).startswith("Unexpected library path:"))

    def test_parse_non_indented_line(self):
        """Lines without leading indentation, at this point in processing, are
        informational.
        """
        # https://github.com/Parquery/pylddwrap/pull/14
        line = (
            "qt/6.0.2/gcc_64/plugins/sqldrivers/libqsqlpsql.so:" +
            " /lib/x86_64-linux-gnu/libpq.so.5:" +
            " no version information available" +
            " (required by qt/6.0.2/gcc_64/plugins/sqldrivers/libqsqlpsql.so)")
        result = lddwrap._parse_line(line=line)  # pylint: disable=protected-access

        self.assertIsNone(result)

    def test_parse_static_with_two_line_output(self) -> None:
        """Test parsing of the output when we ldd a static library."""
        # pylint: disable=protected-access
        deps = lddwrap._cmd_output_parser("\n".join([
            "my_static_lib.so:",
            "    statically linked",
            "",
        ]))

        self.assertListEqual([], deps)

    def test_parse_static_with_single_line_output(self) -> None:
        """Test parsing of the output when we ldd a static library."""
        # pylint: disable=protected-access
        deps = lddwrap._cmd_output_parser("\tstatically linked\n")

        self.assertListEqual([], deps)


class TestAgainstMockLdd(unittest.TestCase):
    def test_pwd(self):
        """Test parsing the captured output  of ``ldd`` on ``/bin/pwd``."""

        with tests.MockLdd(
                out="\n".join([
                    "\tlinux-vdso.so.1 (0x00007ffe0953f000)",
                    "\tlibc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fd548353000)",  # pylint: disable=C0301  # pylint: disable=C0301
                    "\t/lib64/ld-linux-x86-64.so.2 (0x00007fd54894d000)",
                    "",
                ]),
                out_unused=''):
            deps = lddwrap.list_dependencies(
                path=pathlib.Path(PWD), unused=False)

            expected_deps = [
                lddwrap.Dependency(
                    soname="linux-vdso.so.1",
                    path=None,
                    found=True,
                    mem_address="0x00007ffe0953f000",
                    unused=None),
                lddwrap.Dependency(
                    soname='libc.so.6',
                    path=pathlib.Path("/lib/x86_64-linux-gnu/libc.so.6"),
                    found=True,
                    mem_address="0x00007fd548353000",
                    unused=None),
                lddwrap.Dependency(
                    soname=None,
                    path=pathlib.Path("/lib64/ld-linux-x86-64.so.2"),
                    found=True,
                    mem_address="0x00007fd54894d000",
                    unused=None)
            ]

            self.assertEqual(len(expected_deps), len(deps))

            for i, (dep, expected_dep) in enumerate(zip(deps, expected_deps)):
                self.assertListEqual([],
                                     diff_dependencies(
                                         ours=dep, theirs=expected_dep),
                                     "Mismatch at the dependency {}".format(i))

    def test_bin_dir(self):
        """Test parsing the captured output  of ``ldd`` on ``/bin/dir``."""

        # pylint: disable=line-too-long
        with tests.MockLdd(
                out="\n".join([
                    "\tlinux-vdso.so.1 (0x00007ffd66ce2000)",
                    "\tlibselinux.so.1 => /lib/x86_64-linux-gnu/libselinux.so.1 (0x00007f72b88fc000)",
                    "\tlibc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f72b850b000)",
                    "\tlibpcre.so.3 => /lib/x86_64-linux-gnu/libpcre.so.3 (0x00007f72b8299000)",
                    "\tlibdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f72b8095000)",
                    "\t/lib64/ld-linux-x86-64.so.2 (0x00007f72b8d46000)",
                    "\tlibpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f72b7e76000)",
                    "",
                ]),
                out_unused=''):
            # pylint: enable=line-too-long
            deps = lddwrap.list_dependencies(
                path=pathlib.Path(DIR), unused=False)

            expected_deps = [
                lddwrap.Dependency(
                    soname="linux-vdso.so.1",
                    path=None,
                    found=True,
                    mem_address="0x00007ffd66ce2000",
                    unused=None),
                lddwrap.Dependency(
                    soname="libselinux.so.1",
                    path=pathlib.Path("/lib/x86_64-linux-gnu/libselinux.so.1"),
                    found=True,
                    mem_address="0x00007f72b88fc000",
                    unused=None),
                lddwrap.Dependency(
                    soname="libc.so.6",
                    path=pathlib.Path("/lib/x86_64-linux-gnu/libc.so.6"),
                    found=True,
                    mem_address="0x00007f72b850b000",
                    unused=None),
                lddwrap.Dependency(
                    soname="libpcre.so.3",
                    path=pathlib.Path("/lib/x86_64-linux-gnu/libpcre.so.3"),
                    found=True,
                    mem_address="0x00007f72b8299000",
                    unused=None),
                lddwrap.Dependency(
                    soname="libdl.so.2",
                    path=pathlib.Path("/lib/x86_64-linux-gnu/libdl.so.2"),
                    found=True,
                    mem_address="0x00007f72b8095000",
                    unused=None),
                lddwrap.Dependency(
                    soname=None,
                    path=pathlib.Path("/lib64/ld-linux-x86-64.so.2"),
                    found=True,
                    mem_address="0x00007f72b8d46000",
                    unused=None),
                lddwrap.Dependency(
                    soname="libpthread.so.0",
                    path=pathlib.Path("/lib/x86_64-linux-gnu/libpthread.so.0"),
                    found=True,
                    mem_address="0x00007f72b7e76000",
                    unused=None),
            ]

            self.assertEqual(len(expected_deps), len(deps))

            for i, (dep, expected_dep) in enumerate(zip(deps, expected_deps)):
                self.assertListEqual([],
                                     diff_dependencies(
                                         ours=dep, theirs=expected_dep),
                                     "Mismatch at the dependency {}".format(i))

    def test_bin_dir_with_empty_unused(self):
        # pylint: disable=line-too-long
        with tests.MockLdd(
                out="\n".join([
                    "\tlinux-vdso.so.1 (0x00007ffd66ce2000)",
                    "\tlibselinux.so.1 => /lib/x86_64-linux-gnu/libselinux.so.1 (0x00007f72b88fc000)",
                    "\tlibc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f72b850b000)",
                    "\tlibpcre.so.3 => /lib/x86_64-linux-gnu/libpcre.so.3 (0x00007f72b8299000)",
                    "\tlibdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f72b8095000)",
                    "\t/lib64/ld-linux-x86-64.so.2 (0x00007f72b8d46000)",
                    "\tlibpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f72b7e76000)",
                    "",
                ]),
                out_unused=''):
            # pylint: enable=line-too-long
            deps = lddwrap.list_dependencies(
                path=pathlib.Path(DIR), unused=True)

            unused = [dep for dep in deps if dep.unused]
            self.assertListEqual([], unused)

    def test_with_fantasy_unused(self):
        """Test against a fantasy executable with fantasy unused."""
        # pylint: disable=line-too-long
        with tests.MockLdd(
                out="\n".join([
                    "\tlinux-vdso.so.1 (0x00007ffd66ce2000)",
                    "\tlibm.so.6 => /lib64/libm.so.6 (0x00007f72b7e76000)",
                    "",
                ]),
                out_unused="\n".join([
                    "Unused direct dependencies:",
                    "\t/lib64/libm.so.6",
                    "",
                ]),
        ):
            # pylint: enable=line-too-long
            deps = lddwrap.list_dependencies(
                path=pathlib.Path(DIR), unused=True)

            unused = [dep for dep in deps if dep.unused]

            expected_unused = [
                lddwrap.Dependency(
                    soname="libm.so.6",
                    path=pathlib.Path("/lib64/libm.so.6"),
                    found=True,
                    mem_address="0x00007f72b7e76000",
                    unused=True)
            ]

            self.assertEqual(len(expected_unused), len(unused))

            for i, (dep, exp_dep) in enumerate(zip(unused, expected_unused)):
                self.assertListEqual(
                    [], diff_dependencies(ours=dep, theirs=exp_dep),
                    "Mismatch at the unused dependency {}".format(i))

    def test_with_static_library(self) -> None:
        """Test against a fantasy static library."""
        with tempfile.TemporaryDirectory() as tmp_dir:
            lib_pth = pathlib.Path(tmp_dir) / "my_static_lib.so"
            lib_pth.write_text("totally static!")

            with tests.MockLdd(
                    out="\n".join([
                        "my_static_lib.so:",
                        "\tstatically linked",
                        "",
                    ]),
                    out_unused=''):
                # pylint: enable=line-too-long
                deps = lddwrap.list_dependencies(path=lib_pth, unused=True)

                # The dependencies are empty since the library is
                # statically linked.
                self.assertListEqual([], deps)


class TestSorting(unittest.TestCase):
    def test_sorting_by_all_attributes(self) -> None:
        # pylint: disable=line-too-long
        with tests.MockLdd(
                out="\n".join([
                    "\tlinux-vdso.so.1 (0x00007ffd66ce2000)",
                    "\tlibselinux.so.1 => /lib/x86_64-linux-gnu/libselinux.so.1 (0x00007f72b88fc000)",
                    "\tlibc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f72b850b000)",
                    "\tlibpcre.so.3 => /lib/x86_64-linux-gnu/libpcre.so.3 (0x00007f72b8299000)",
                    "\tlibdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f72b8095000)",
                    "\t/lib64/ld-linux-x86-64.so.2 (0x00007f72b8d46000)",
                    "\tlibpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f72b7e76000)",
                    "",
                ]),
                out_unused=''):
            # pylint: enable=line-too-long

            for attr in lddwrap.DEPENDENCY_ATTRIBUTES:
                deps = lddwrap.list_dependencies(
                    path=pathlib.Path(DIR), unused=True)

                # pylint: disable=protected-access
                lddwrap._sort_dependencies_in_place(deps=deps, sort_by=attr)

                previous = getattr(deps[0], attr)
                previous = '' if previous is None else str(previous)

                for i in range(1, len(deps)):
                    current = getattr(deps[i], attr)
                    current = '' if current is None else str(current)

                    self.assertLessEqual(
                        previous, current,
                        ("The dependencies must be sorted according to "
                         "attribute {!r}: {!r}").format(attr, deps))


if __name__ == '__main__':
    unittest.main()


precommit.py
#!/usr/bin/env python3
"""Runs precommit checks on the repository."""
import argparse
import os
import pathlib
import shutil
import subprocess
import sys
import sysconfig
import textwrap

# os.fspath() is 3.6+.  str is a reasonable substitute
# https://docs.python.org/3.9/library/os.html#os.fspath
fspath = getattr(os, "fspath", str)


class ToxNotFoundError(Exception):
    """Rasie when tox not found in expected locations."""


def main() -> int:
    """Execute the main routine."""
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--overwrite",
        help=
        "Overwrites the unformatted source files with the well-formatted code in place. "  # pylint: disable=line-too-long
        "If not set, an exception is raised if any of the files do not conform to the style guide.",  # pylint: disable=line-too-long
        action='store_true')

    args = parser.parse_args()

    tox_environments = ["check", "_auto_version"]

    if args.overwrite:
        tox_environments.insert(0, "format")

    tox_from_path = shutil.which("tox")

    maybe_scripts = sysconfig.get_path("scripts")
    tox_from_scripts = pathlib.Path(maybe_scripts).joinpath(
        "tox") if maybe_scripts is not None else None

    if tox_from_path is not None:
        tox = pathlib.Path(tox_from_path)
    elif tox_from_scripts is not None and tox_from_scripts.is_file():
        tox = tox_from_scripts
    else:
        message = textwrap.dedent(f'''\
        'tox' executable not found in PATH or scripts directory
            PATH: {os.environ['PATH']}
            scripts: {maybe_scripts}
        ''')
        raise ToxNotFoundError(message)

    repo_root = pathlib.Path(__file__).parent

    tox_ini = repo_root.joinpath("tox.ini")

    command = [
        fspath(tox), "-c",
        fspath(tox_ini), "-e", ",".join(tox_environments)
    ]
    completed_process = subprocess.run(command)  # pylint: disable=W1510

    return completed_process.returncode


if __name__ == "__main__":
    sys.exit(main())


lddwrap/__init__.py
#!/usr/bin/env python3
"""Wrap ldd *nix utility to determine shared libraries required by a program."""

import collections
import copy
import json
import os
import pathlib
import re
import subprocess
# pylint: disable=unused-import
from typing import Any, Dict, List, Mapping, Optional, TextIO

import icontract

DEPENDENCY_ATTRIBUTES = ['soname', 'path', 'found', 'mem_address', 'unused']


# yapf: disable
@icontract.invariant(
    lambda self: not (self.soname is None and self.found) or (
            self.path is not None and self.mem_address is not None)
)
@icontract.invariant(
    lambda self: not (self.path is None and self.found) or (
            self.soname is not None and self.mem_address is not None)
)
@icontract.invariant(
    lambda self: not (self.mem_address is None and self.found) or (
            self.soname is not None and self.path is not None)
)
# yapf: enable
class Dependency:
    """
    Represent a shared library required by a program.

    :ivar found: True if ``ldd`` could resolve the library
    :vartype found: bool

    :ivar soname: library name
    :vartype soname: Optional[str]

    :ivar path: path to the library
    :vartype path: Optional[pathlib.Path]

    :ivar mem_address: hex memory location
    :vartype mem_address: Optional[str]

    :ivar unused: library not used
    :vartype unused: Optional[bool]
    """

    @icontract.ensure(
        lambda self: all(hasattr(self, col) for col in DEPENDENCY_ATTRIBUTES),
        "All expected attributes are present so that "
        "lists of dependencies can be sorted dynamically")
    def __init__(self,
                 found: bool,
                 soname: Optional[str] = None,
                 path: Optional[pathlib.Path] = None,
                 mem_address: Optional[str] = None,
                 unused: Optional[bool] = None) -> None:
        """Initialize the dependency with the given values."""
        # pylint: disable= too-many-arguments
        self.soname = soname
        self.path = path
        self.found = found
        self.mem_address = mem_address
        self.unused = unused

    def __str__(self) -> str:
        """Transform the dependency to a human-readable format."""
        return "soname: {}, path: {}, found: {}, mem_address: {}, unused: {}" \
               "".format(self.soname, self.path, self.found,
                         self.mem_address, self.unused)

    # pylint: disable=unsubscriptable-object
    def as_mapping(self) -> Mapping[str, Any]:
        """
        Transform the dependency to a mapping.

        Can be converted to JSON and similar formats.
        """
        return collections.OrderedDict([("soname", self.soname),
                                        ("path", str(self.path)),
                                        ("found", self.found),
                                        ("mem_address", self.mem_address),
                                        ("unused", self.unused)])


_LDD_ARROW_OUTPUT_RE = re.compile(
    r"(?P<soname>.+)\s=>\s(?P<dep_path>.*)\s\(?(?P<mem_address>\w*)\)?")
_LDD_NON_ARROW_OUTPUT_RE = re.compile(
    r"(?P<dep_path>.+)\s\(?(?P<mem_address>\w*)\)?")


# pylint: disable=too-many-branches
def _parse_line(line: str) -> Optional[Dependency]:
    """
    Parse single line of ldd output.

    :param line: to parse
    :return: dependency or None if line was empty

    """
    # pylint: disable=line-too-long
    # There are two types of outputs for a dependency, with or without soname.
    # The VDSO is a special case (see https://man7.org/linux/man-pages/man7/vdso.7.html)
    #
    # For example:
    # VDSO (Ubuntu 16.04): linux-vdso.so.1 =>  (0x00007ffd7c7fd000)
    # VDSO (Ubuntu 18.04): linux-vdso.so.1 (0x00007ffe2f993000)
    # with soname: 'libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f9a19d8a000)'
    # without soname: '/lib64/ld-linux-x86-64.so.2 (0x00007f9a1a329000)'
    # with soname but not found: 'libboost_program_options.so.1.62.0 => not found'
    # with soname but without rpath: 'linux-vdso.so.1 =>  (0x00007ffd7c7fd000)'
    # pylint: enable=line-too-long
    found = not 'not found' in line
    soname = None
    dep_path = None
    mem_address = None
    if '=>' in line:
        mtch = _LDD_ARROW_OUTPUT_RE.match(line)
        if not mtch:
            raise RuntimeError(("Unexpected ldd output. Expected to match {}, "
                                "but got: {!r}").format(
                                    _LDD_ARROW_OUTPUT_RE.pattern, line))
        if found:
            soname = mtch["soname"]
            if mtch["dep_path"]:
                dep_path = pathlib.Path(mtch["dep_path"])
            if mtch["mem_address"]:
                mem_address = mtch["mem_address"]
        else:
            if os.sep in mtch["soname"]:
                # This is a special case where the dep_path comes before the
                # arrow and we have no soname
                dep_path = pathlib.Path(mtch["soname"])
            else:
                soname = mtch["soname"]
    else:
        # Please see https://github.com/Parquery/pylddwrap/pull/14
        if 'no version information available' in line:
            return None

        mtch = _LDD_NON_ARROW_OUTPUT_RE.match(line)
        if not mtch:
            raise RuntimeError(("Unexpected ldd output. Expected to match {}, "
                                "but got: {!r}").format(
                                    _LDD_NON_ARROW_OUTPUT_RE.pattern, line))
        # Special case for linux-vdso
        if mtch["dep_path"].startswith("linux-vdso"):
            soname = mtch["dep_path"]
        else:
            dep_path = pathlib.Path(mtch["dep_path"])

        found = True
        mem_address = mtch["mem_address"]

    # Sanity check to see if it didn't parse garbage:
    # dep_path should have at least a `/` somewhere in the filepath
    if dep_path and os.sep not in str(dep_path):
        raise RuntimeError("Unexpected library path: {}".format(dep_path))

    return Dependency(
        soname=soname, path=dep_path, found=found, mem_address=mem_address)


@icontract.require(lambda path: path.is_file())
def list_dependencies(path: pathlib.Path,
                      unused: bool = False,
                      env: Optional[Dict[str, str]] = None) -> List[Dependency]:
    """
    Retrieve a list of dependencies of the given binary.

    >>> import shutil
    >>> ls = shutil.which("ls")
    >>> path = pathlib.Path(ls)
    >>> deps = list_dependencies(path=path)
    >>> deps[0].soname
    'linux-vdso.so.1'

    :param path: path to a file
    :param unused:
        if set, check if dependencies are actually used by the program
    :param env:
        the environment to use.

        If ``env`` is None, currently active env will be used.
        Otherwise specified env is used.
    :return: list of dependencies
    """
    # We need to use /usr/bin/env since Popen ignores the PATH,
    # see https://stackoverflow.com/questions/5658622
    proc = subprocess.Popen(
        ["/usr/bin/env", "ldd", path.as_posix()],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        universal_newlines=True,
        env=env)

    out, err = proc.communicate()
    if proc.returncode != 0:
        raise RuntimeError(
            "Failed to ldd external libraries of {} with code {}:\nout:\n{}\n\n"
            "err:\n{}".format(path, proc.returncode, out, err))

    dependencies = _cmd_output_parser(cmd_out=out)  # type: List[Dependency]

    if unused:
        proc_unused = subprocess.Popen(
            ["/usr/bin/env", "ldd", "--unused",
             path.as_posix()],
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            universal_newlines=True,
            env=env)

        out_unused, err_unused = proc_unused.communicate()
        # return code = 0 -> no unused dependencies,
        # return code = 1 -> some unused dependencies
        if proc_unused.returncode not in [0, 1]:
            raise RuntimeError(
                "Failed to ldd external libraries of {} with code {}:\nout:\n"
                "{}\n\nerr:\n{}".format(path, proc.returncode, out_unused,
                                        err_unused))

        dependencies = _update_unused(
            dependencies=dependencies, out_unused=out_unused)

    return dependencies


def _cmd_output_parser(cmd_out: str) -> List[Dependency]:
    """
    Parse the command line output.

    :param cmd_out: command line output
    :return: List of dependencies
    """
    dependencies = []  # type: List[Dependency]

    lines = [line.strip() for line in cmd_out.split('\n') if line.strip() != '']

    if len(lines) == 0:
        return []

    # Static libraries can appear in multiple forms:
    # - a first line refering to the library and a second line indicating
    #   that the library was statically linked,
    # - only a single line indicating statically linked.
    # See: https://github.com/Parquery/pylddwrap/issues/12
    if 1 <= len(lines) <= 2 and lines[-1] == 'statically linked':
        return []

    for line in lines:
        dep = _parse_line(line=line)
        if dep is not None:
            dependencies.append(dep)

    return dependencies


def _update_unused(dependencies: List[Dependency],
                   out_unused: str) -> List[Dependency]:
    """
    Set "unused" property of the dependencies.

    Updates the "unused" property of the dependencies using the output string
    from ldd command.

    :param dependencies: List of dependencies
    :param out_unused: output from command ldd --unused
    :return: updated list of dependencies
    """
    unused_dependencies = []  # type: List[pathlib.Path]

    for line in [
            line.strip() for line in out_unused.split('\n')
            if line.strip() != ''
    ]:
        # skip first line because it's no dependency
        if line != "Unused direct dependencies:":
            unused_dependencies.append(pathlib.Path(line.strip()))

    for dep in dependencies:
        dep.unused = dep.path in unused_dependencies

    return dependencies


# pylint: disable=unnecessary-lambda
@icontract.require(lambda sort_by: sort_by in DEPENDENCY_ATTRIBUTES)
@icontract.snapshot(
    lambda deps: copy.copy(deps), name='deps', enabled=icontract.SLOW)
@icontract.ensure(
    lambda deps, OLD: set(deps) == set(OLD.deps), enabled=icontract.SLOW)
# pylint: enable=line-too-long
def _sort_dependencies_in_place(deps: List[Dependency], sort_by: str) -> None:
    """Order the dependencies by the given ``sort_by`` attribute."""
    if len(deps) < 2:
        return

    def compute_key(dep: Dependency) -> str:
        """Produce key for the sort."""
        assert hasattr(dep, sort_by)
        key = getattr(dep, sort_by)

        assert key is None or isinstance(key, (bool, str, pathlib.Path))

        if key is None:
            return ''

        return str(key)

    deps.sort(key=compute_key)


def _output_verbose(deps: List[Dependency], stream: TextIO) -> None:
    """
    Output dependencies in verbose, human-readable format to the ``stream``.

    :param deps: list of dependencies
    :param stream: output stream
    :return:
    """
    table = [["soname", "path", "found", "mem_address", "unused"]]

    for dep in deps:
        table.append([
            str(dep.soname),
            str(dep.path),
            str(dep.found),
            str(dep.mem_address),
            str(dep.unused)
        ])

    stream.write(_format_table(table))


def _output_json(deps: List[Dependency], stream: TextIO) -> None:
    """
    Output dependencies in a JSON format to the ``stream``.

    :param deps: list of dependencies
    :param stream: output stream
    :return:
    """
    json.dump(obj=[dep.as_mapping() for dep in deps], fp=stream, indent=2)


@icontract.require(
    lambda table: not table or all(len(row) == len(table[0]) for row in table))
@icontract.ensure(lambda table, result: result == "" if not table else True)
@icontract.ensure(lambda result: not result.endswith("\n"))
def _format_table(table: List[List[str]]) -> str:
    """
    Format the table as equal-spaced columns.

    :param table: rows of cells
    :return: table as string
    """
    cols = len(table[0])

    col_widths = [max(len(row[i]) for row in table) for i in range(cols)]

    lines = []  # type: List[str]
    for i, row in enumerate(table):
        parts = []  # type: List[str]

        for cell, width in zip(row, col_widths):
            parts.append(cell.ljust(width))

        line = " | ".join(parts)
        lines.append(line)

        if i == 0:
            border = []  # type: List[str]

            for width in col_widths:
                border.append("-" * width)

            lines.append("-+-".join(border))

    result = "\n".join(lines)

    return result
